crates/re_query_cache/src/flat_vec_deque.rs

```diff
@@ -0,0 +1,789 @@
+        FlatVecDeque::<T>::remove_range(self, range);
+    }
+
+    #[inline]
+    fn dyn_truncate(&mut self, at: usize) {
+        FlatVecDeque::<T>::truncate(self, at);
+    }
+}
+
+// ---
+
+/// A double-ended queue implemented with a pair of growable ring buffers, where every single
+/// entry is a flattened array of values.
+///
+/// You can think of this as the native/deserialized version of an Arrow `ListArray`.
+///
+/// This is particularly when working with many small arrays of data (e.g. Rerun's `TimeSeriesScalar`s).
+//
+// TODO(cmc): We could even use a bitmap for T=Option<Something>, which would bring this that much
+// closer to a deserialized version of an Arrow array.
+#[derive(Debug, Clone)]
+pub struct FlatVecDeque<T> {
+    values: VecDeque<T>,
+    offsets: VecDeque<usize>,
+}
+
+impl<T: SizeBytes> SizeBytes for FlatVecDeque<T> {
+    #[inline]
+    fn heap_size_bytes(&self) -> u64 {
+        // NOTE: It's all on the heap at this point.
+
+        let values_size_bytes = if T::is_pod() {
+            (self.num_values() * std::mem::size_of::<T>()) as _
+        } else {
+            self.values
+                .iter()
+                .map(SizeBytes::total_size_bytes)
+                .sum::<u64>()
+        };
+
+        let offsets_size_bytes = self.num_entries() * std::mem::size_of::<usize>();
+
+        values_size_bytes + offsets_size_bytes as u64
+    }
+}
+
+impl<T> Default for FlatVecDeque<T> {
+    #[inline]
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T> FlatVecDeque<T> {
+    #[inline]
+    pub const fn new() -> Self {
+        Self {
+            values: VecDeque::new(),
+            offsets: VecDeque::new(),
+        }
+    }
+
+    #[inline]
+    pub fn from_vecs(entries: impl IntoIterator<Item = Vec<T>>) -> Self {
+        let mut this = Self::new();
+
+        // NOTE: Do not use any of the insertion methods, they rely on `from_vecs` in the first
+        // place!
+        let mut value_offset = 0;
+        for entry in entries {
+            value_offset += entry.len(); // increment first!
+            this.offsets.push_back(value_offset);
+            this.values.extend(entry);
+        }
+
+        this
+    }
+
+    /// How many entries are there in the deque?
+    ///
+    /// Keep in mind: each entry is itself an array of values.
+    /// Use [`Self::num_values`] to get the total number of values across all entries.
+    #[inline]
+    pub fn num_entries(&self) -> usize {
+        self.offsets.len()
+    }
+
+    /// How many values are there in the deque?
+    ///
+    /// Keep in mind: each entry in the deque holds an array of values.
+    /// Use [`Self::num_entries`] to get the total number of entries, irrelevant of how many
+    /// values each entry holds.
+    #[inline]
+    pub fn num_values(&self) -> usize {
+        self.values.len()
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.num_entries() == 0
+    }
```

Comment:
This needs a clarifying docstring. Is `Self::from_vecs([vec![]])` empty or not?

I would suggest skipping `is_empty` completely to avoid the ambiguity. Users can always `if num_values() == 0` or `num_entries() == 0` with no loss of performance